Add features spec for Resource extensibility (#58)
radius-project/design-notes: resources/2024-06-resource-extensibility-feature-spec.md

- Assumption: Service providers of proprietary services will find value in creating and publishing Recipes for their services hoping to increase the adoption of their services.

## Current state 
<!-- If we already have some ongoing investment in this area, summarize the current state and point to any relevant documents. -->

### Current state in Radius

[Radius Extenders](https://docs.radapp.io/guides/author-apps/custom/overview/#extenders) enables users to author custom resource types but they are weakly typed. They allow you to pass in any property or secret and for cases where the developer/operator need to extend and reference a simple resource type definition. Extenders are for untyped/ weakly typed resources and do not provide a lot of customization for users wanting to have guardrails in their platforms with strongly typed resource definitions following a strict set of rules and restrictions. 

Below is an example extender implementation for PostgreSQL

``` bicep
extension radius

resource extender 'Applications.Core/extenders@2023-10-01-preview' = {
  name: 'postgresql'
  properties: {
    name : 'postgresql'
    sku : 'standard'
    }
  }
}
```

> The properties in the extender resource are free form key value pairs. The user can pass in any text as SKU as there are no validation. In a mature enterprise, platform engineers or IT operators want to control the SKU a developer can pass in and provision. They would allow a set of SKU's that are allowed and restrict the user from passing in any other SKU. User defined types will enable the platform engineer to define the resource type definition in a strong typed way with guardrails and validation rules adhering to their organizational policies and compliance. 

Author: ```suggestion
The properties in the extender resource are free form key value pairs. The user can pass in any text as SKU as there are no validation. In a mature enterprise, platform engineers or IT operators want to control the SKU a developer can pass in and provision. They would allow a set of SKU's that are allowed and restrict the user from passing in any other SKU. User defined types will enable the platform engineer to define the resource type definition in a strong typed way with guardrails and validation rules adhering to their organizational policies. 
```

# Resource Extensibility (User Defined Resource Types)

* **Status**: Pending
* **Author**: Reshma abdul Rahim (@reshrahim)

## Summary

The cloud native landscape is changing and expanding at a rapid pace like never before. Many enterprises use a wide range of technologies together for achieving their cloud-native strategy.  For any technology that’s a newcomer and helps solve a problem in the cloud native landscape, users look for an easy and seamless way to integrate their existing tools and technologies and incrementally adopt the new technology to their strategy. For Radius, we have heard requests from our users/community to support technologies that their applications are tightly coupled with E.g.: an internal messaging service or a technology they absolutely love E.g.: PostgreSQL/Kafka.

We need to enable the open-source community to build and experiment with imperfect things. We need to provide an extensibility model that supports “Bring your own technology”, define and use it with Radius. This will help us to meet the community where they are and enable them to experiment and leverage their work as open-source contributions.

> One of the high value extensibility points in Radius is Recipes. We have received interests to create custom resource types, define Recipes for the custom resource types and use it in the Radius application. Today Radius Extenders helps in creating custom resource types, but they are untyped and have limitations. The goal of providing resource extensibility is to empower developers or infrastructure operators to author and run their applications with custom resource types seamlessly in Radius and use all the other features such as Recipes, connections and app graph with ease and flexibility.

Reviewer: ```suggestion
One of the high value extensibility points in Radius is Recipes. We have received interests to create custom resource types, define Recipes for the custom resource types and use it in the Radius application. Today Radius Extenders helps unblock users by modeling any kind of technology in an untyped way, but they are not first-class in the same way a custom type is. The goal of providing resource extensibility is to empower developers or infrastructure operators to author and run their applications with custom resource types seamlessly in Radius and use all the other features such as Recipes, connections and app graph with ease and flexibility.
```

nit: extenders aren't custom types. They are the "unknown" type.

> One of the high value extensibility points in Radius is Recipes. We have received interests to create custom resource types, define Recipes for the custom resource types and use it in the Radius application. Today Radius Extenders helps in creating custom resource types, but they are untyped and have limitations. The goal of providing resource extensibility is to empower developers or infrastructure operators to author and run their applications with custom resource types seamlessly in Radius and use all the other features such as Recipes, connections and app graph with ease and flexibility.

Reviewer 2: > they are untyped and have limitations

I recommend elaborating a bit more on specific limitations and why extenders might not be a workable substitute

### Top level goals

1. Enable users to author their applications with custom resource types in Radius without having to write Go code to integrate with Radius components and deploy their application seamlessly.
2. Automatically enable Radius features such as Recipes, Connections, App graph, rad CLI for custom resource types.
3. Enable users to contribute and open-source the custom resource types and recipes to the community. This will accelerate the adoption of Radius across the cloud native community.
4. Radius uses the resource extensibility framework to refactor its already built in resource types.

### Non-goals (out of scope)

> We will not cover the design and user experience for "Recipes for any resource" in this document to keep the scope of the work simple.

Reviewer 3: Are there dependencies between these two features? Wondering if recipes for any resource would be needed to deploy custom resource types

Author: A user can use UDT's in the place of `Recipes for any resource` but they need to manage the resource type definition. With Recipes for any resource, the user doesn't have to manage the schema definitions and simply create recipes for any Radius type.

> ## User profile and challenges

Reviewer 2: Recommend adding a bit more around challenges of the *current state* - what are users missing with `extenders` today, etc.

Author: Expanded on extenders below

**Open-source community**: Building a sustainable open-source community is crucial to the success of any open-source project, including Radius. We need to cultivate an ecosystem for the open-source community to thrive, innovate and produce high quality work. Users from the community are motivated to contribute for different reasons:
    -  They want to use the project at their work and need a feature, 
    -  The project triggered their interests in the cloud native landscape and want to use the technology.
    -  They want to code in a particular language E.g. :Go
    -  They want to integrate their existing tools and technologies with the project

> Today Radius enables users to get started on contributing to Radius with good-first-issues but doesn’t have a model to promote contributions further. Since the beginning of its open source launch, Dapr created the components repository and invited users to build their own components to unblock their scenarios and in turn folks contributed to the project. Learning from Dapr, Radius needs to have the extensibility points defined for contributors to interact with the project so that they can build out the things that they need to enable their scenarios and in turn the community gets benefitted with those contributions.

Reviewer 2: ```suggestion
Today Radius enables users to get started on contributing to Radius with good-first-issues but doesn’t have a model to promote contributions further. Since the beginning of its open source launch, [Dapr](www.dapr.io) created the components repository and invited users to build their own components to unblock their scenarios and in turn folks contributed to the project. Learning from Dapr, Radius needs to have the extensibility points defined for contributors to interact with the project so that they can build out the things that they need to enable their scenarios and in turn the community gets benefitted with those contributions.
```

- IT Operators: IT operators are responsible for managing the infrastructure and ensuring that the applications are running smoothly. They are responsible for maintaining the infrastructure and providing support for the infrastructure. They are the primary users of Radius Recipes as they are responsible for defining the infrastructure-as-code templates for the applications.

- System Reliability Engineers (SREs) : SREs are responsible for ensuring the reliability of the applications and services that are deployed in the cloud. They are responsible for maintaining the infrastructure and ensuring that the applications are running smoothly. They maintain the infrastructure-as-code templates and provide support for the applications that are deployed. 

- System integrators : System integrators are partners who help enterprises integrate proprietary services with cloud native applications. They are responsible for designing and building cloud native applications on behalf of their customers.

- Developers : Developers are responsible for building the cloud native applications. They are responsible for writing the code, designing and maintaining the applications.

> - Open-source contributors : Cloud native open-source contributors can be any of the above personas who are interested in contributing to the cloud native projects. 

Reviewer 2: They may also be interested in leveraging Radius to make adoption of their own technologies easier (e.g. folks at Postgres, etc.)

Author: review feedback : add providers of solutions to system integrators persona

### Positive user outcomes

- Author and deploy : I can bring my own applications with custom resources/services and integrate with Radius seamlessly. I don’t have to write Go code to integrate with Radius components and can use simple spec to generate the resource definition and integrate with Radius 

- Recipes for user defined types: I can create Recipes for the custom resource types and deploy via Radius 

> - OSS contributions: I can create and contribute Recipes for the custom resource types for the community to use.

Reviewer 4: Will the custom resource types be merged into the main repository of Radius, or will we have a separate repository for this?

Author: this will be a separate repository with e-2-e validation

## Key scenarios

### Scenario 1: Deb integrates the Budgets app with Radius

Deb, a platform engineer at a Financial Services company, wants to integrate the Budgets app with Radius. The Budgets app relies on an internal messaging service called Plaid. Deb wants to use Plaid as a resource type in Radius to deploy the application seamlessly. He needs a way to define and use Plaid as a custom resource type in Radius without having to write Go code. This will enable him to leverage all the features of Radius, such as Recipes, Connections, and the rad CLI, with ease and flexibility.

### Scenario 2: Amy contributes and open-sources PostgreSQL support to Radius

Amy is a system integrator who helps customers build cloud native applications on AWS. Amy heard about Radius and starts integrating their customer workloads in Radius. She sees a lot of her scenarios involve PostgreSQL and finds Radius doesn’t support that yet. She wants to contribute the PostgreSQL support to Radius and open source it so anyone can use it

### Scenario 3: Raj publishes Recipes for proprietary services

> Raj is a partner or a system integrator who helps enterprises integrate proprietary service for eg : Oracle database in their cloud native applications. Raj tries out Radius and wants to write Recipes to create an Oracle Database with a standard set of best practices and policies integrated. He works with a lot of customers who are wanting to use the Oracle Database recipe in their applications.

Reviewer 2: > He works with a lot of customers who are wanting to use the Oracle Database recipe in their applications.
 
I think this scenario might be even more illustrative if Raj isn't just an integrator, but perhaps someone who works on the Oracle Database team and is looking to expand their adopter/user base. By contributing an Oracle Database recipe, they are hoping to lower the barrier for entry and attract Radius users to using Oracle products. This would give this scenario a "selfishly contributes to Radius" story where open-source and enterprise collaborate for a win-win situation.

Author: flush some more details aeround soultion provider on the benefits

## Key dependencies and risks
<!-- What dependencies must we take in order to enable this scenario? -->

<!-- What other risks are you aware of that need to be mitigated. If you have a mitigation in mind, summarize here. -->
<!-- **Dependency Name** – summary of dependency.  Issues/concerns/risks with this dependency -->
<!-- **Risk Name** – summary of risk.  Mitigation plan if known. If it is not yet known, no problem. --> 

- Dependency: Bicep compiler merge support
    - We need to ensure that the ongoing Bicep compiler merge work is complete so that custom resource types can be dynamically generated for the user defined resource types.
    - Mitigation: Work closely with the Bicep team to ensure that the merge is completed on time and that the necessary features are available to enable custom resource types in Radius .

- Risk: Adoption and community engagement

>     - There is a risk that users may not fully embrace the extensibility feature or contribute custom resource types to the community.

Reviewer 4: We can add a few custom resource types and a tutorial to show external contributors how they might do it.

Reviewer 4: Are we going to be requiring functional tests for all the custom resource types added? That might have some security issues in the pipelines we have. Like secrets and stuff.

Author: flush some more details on how we want to make this experience simple and easier to use

Author: Added some details around secrets handling and functional tests in the security risk below.

    - Mitigation: Actively engage with users, provide clear documentation and resources, and foster a supportive and inclusive community to encourage adoption and contributions.

> - Risk: Compatibility and maintainability

Reviewer 3: Is the expectation that the author of the new type will maintain it? Or is this work that Radius picks up once it gets accepted?

Author: We need to think through the experience from both the types getting added to Radius and non-Radius types. More details to be covered in a separate doc focussed on contributions

- Risk: Security and trust
    - Allowing users to contribute custom resource types introduces potential security risks and the need to establish trust in the community-contributed code.
    - Mitigation: Implement strict security measures, such as code reviews and vulnerability scanning, and establish a transparent review process for community contributions.

## Key assumptions to test and questions to answer
<!-- If you are making assumptions that, if incorrect, would cause us to significantly alter our approach to this scenario, make them explicit here.  Also call out how / when you plan to validate key assumptions. -->
<!-- What big questions must we answer in order to clarify our plan for this scenario.  When and how do you plan to answer those questions (prototype feature x, user research, competitive research, etc) -->

- Assumption: Users will find value in the extensibility feature and actively contribute to the community.
- Assumption: Users will use extensibility in lieu of existing extenders because they want to take advantage of things like strong typing and validation in Radius

> 

Reviewer 2: Another assumption might be that providers of proprietary software would be interested to contribute recipes to the Radius community in hopes of driving adoption for their own products.

## Current state 
<!-- If we already have some ongoing investment in this area, summarize the current state and point to any relevant documents. -->

### Existing solutions

| Solution | Description | Challenges |
|----------|-------------|------------|
| Kubernetes CRD's | Kubernetes Custom Resource Definitions (CRDs) are a way to extend the Kubernetes API and create custom resource types. CRDs allow users to define their own resource types and controllers to manage those resources. CRDs are widely used in the Kubernetes ecosystem to create custom resources for various use cases. | Complicated to author CRDs ; Versioning and Maintenance ; Ensuring backward compatibility |
| Crossplane XRD's | Built on top of Kubernetes CRD's  |  |

### Current state in Radius

> [Radius Extenders](https://docs.radapp.io/guides/author-apps/custom/overview/#extenders) enables users to author custom resource types but they are weakly types. They allow you to pass in any property or secret and for cases where the developer/operator need to extend and reference a simple resource type definition. Extenders are for untyped/ weakly typed resources and do not provide a lot of customization for users wanting to have guardrails in their platforms with strongly typed resource definitions following a strict set of rules and restrictions. 

Reviewer 2: ```suggestion
[Radius Extenders](https://docs.radapp.io/guides/author-apps/custom/overview/#extenders) enables users to author custom resource types but they are weakly typed. They allow you to pass in any property or secret and for cases where the developer/operator need to extend and reference a simple resource type definition. Extenders are for untyped/ weakly typed resources and do not provide a lot of customization for users wanting to have guardrails in their platforms with strongly typed resource definitions following a strict set of rules and restrictions. 
```

| Ship the custom platform | Registering the API's to the platform, providing documentation, onboarding and support | intuitive, seamless, collaborative |
| Maintain the custom platform | Versioning and maintaining the API's | Ease of support, sense of ownership (decision making) | 
| Support for compliance and security | Compatibility and security policies | Trust the platform |
| Open-source custom resources | Repository and pipeline to validate, testing the infrastructure and publishing  | Desire to contribute, easy to use, low overhead /dependency to publish |

As an IT operator and developer trying to build and maintain the application, I have the following jobs to be done and requirements to meet:

#### Main job : Use the custom platform to build and maintain the application

|  Related jobs | Functional aspects | Emotional aspects | 
| ------------- | ------------------ | ----------------- | 
| Use the custom app platform to build my application  | Learn, define and consume the APIs | Easy to use, adopt, low concept count | 

> | Manage and deploy the custom resources in the application platform| Create a IaC template for the custom resource type with compliance and other policies | Easy to author, test and deploy |  

Reviewer: I'm still unsure about including recipe authoring as part of this main job (`Use the custom platform to build and maintain the application`).

It feels better to make to include recipe authoring as part of `Build a custom platform for application development` or as its own main job.

Recipes aren't application specific, and aren't a developer responsibility. They are closer to the platform than to any application.

Author: separate the IT operator from the developer job

*Community supported assets meets their requirements* : Enterprises have a wide range of requirements for their applications based on the nature of the application, market, region, hosting models,compliance and security policies. Radius can enable the community to provide and use the resource type definitions and IaC templates that meet the requirements of the enterprises and meet them where they are. This will enable users to build what they want via Radius thus unblocking them to build their applications.

*Maintenance and support for the community supported assets* : Radius can provide the necessary tools and services to maintain and support the community supported assets. This will help the community to contribute, maintain the assets and provide support to the users who are using the assets.

**Solution in Radius**

1. Resource extensibility for defining the custom resource types in Radius
1. Recipes for the custom resource types
1. Community supported resource type definitions
1. Community supported Recipes
1. Radius auto generates AWS, Azure Recipes from the definitions (Recipes for any resource) 

> We will cover the design and user experience for all the above solutions except for "Recipes for any resource" in this document to keep the scope of the work simple.

Reviewer: First off, all of the content in this section is good. This is the right idea and we should do these things. 

----

However, I think all of this content goes with `Specify the contract for the key services in the application platform`. It's supporting work we do to help users specify the contract. My feedback is about the organization of the information and not about the ideas 👍 

In terms of solution - I don't think we can/should contribute much to `Define the set of key technologies for the enterprise applications`. Organizations have typically identified their "stack" independent of Radius, and those decisions are motivated by concerns separate from Radius. I think it's enough for us to mention that here.

Author: Updated the content to reflect the feedback

#### Specify the contract for the key services in the application platform

As a platform engineer, I want to provide the custom platform with the following: 

*Resource schema authoring*: Define a schema for the custom resource types that my application uses in Radius. The schema defines the structure and properties of the custom resource type. 

> *Auto guidance for schema authoring*: The resource schema authoring tool provides auto guidance and validation to help the user author the schema. The auto guidance tool provides suggestions and recommendations to the user based on the schema that they are authoring. 

Reviewer: Would like to hear more about auto-guidance .. is this copilot? is this intellisense? something else?

Author: change this to auto completion for clarity

| Typespec | Developer friendly; Maintenance and upgrades are easier; Errors and guidance violations at development time  | Microsoft developed language; May not be widely adopted by the cloud-native communities ; Yet another language to adopt | 
| OpenAPI YAML/JSON | More common and standard format across the open-source communities ; K8s CRD's follow the structure of Open API  | Managing versions and ensuring backward compatibility is challenging |

*User groups and their needs*

There are two types of users that we need to cater to:
1. Users who are new or comfortable with experimenting a new schema format typespec
1. Users who are familiar with openAPI schema format and are proficient in defining the schema in YAML/JSON format

Hence, we need to support both the schema formats in Radius to support both user groups.

*User experience*

1. Deb is a platform engineer at Contoso open to use typespec and wants to author a custom resource Plaid in Radius

    Pre-requisites: User has typespec compiler/CLI installed

    1. Deb uses the radCLI to scaffold a template typespec definition

        ```bash
        rad resource-provider init contoso --template-kind typespec
        ```

>         A sample plaid.tsp file and manifest files are created in the application folder with all the typespec dependencies installed.

Author: ```suggestion
        A sample contoso.tsp file and manifest files are created in the application folder with all the typespec dependencies installed.
```

                queue:string;
            }
        }

        ```

    1. VSCode Type spec extension warns Deb on the errors and warnings as he types
     
        ![alt text](2024-06-resource-extensbility/errors.png)

    1. Deb uses intellisense to autocomplete the schema and properties

        ![alt text](2024-06-resource-extensbility/intellisense.png)

    1. Deb chooses a version for the schema

        ```
        import "@typespec/http";
        import "@typespec/rest";
        import "@typespec/openapi3";
        import "@typespec/versioning";

        using TypeSpec.Http;
        using TypeSpec.Rest;
        using TypeSpec.Versioning;

        @service({title: "Contoso"})
        @server("http://contoso.plaid.com", "endpoint for plaid service")
        @versioned(Versions)
        namespace Contoso;

        @route("/planes/radius/Contoso.Messaging")
        namespace plaid {
            model plaid {
                name:string;
                description:string;
                queue:string;
            }
        }

        enum Versions {
            v1,
        }
        ```
   1. Deb registers the schema in Radius
    
        ```bash

>         rad resource-provider register contoso --template-kind typespec -e myenv -v v1

Reviewer 4: Will there be a way to register a custom resource type that is on GitHub or on cloud somewhere? Like from the Radius custom resource type library?

Author: This could be definitely something we can think about as part of e-2-e automation

>         rad resource-provider register contoso --template-kind typespec -e myenv -v v1

Reviewer 5: How are we passing in typespec input to this command?

        ```
        Radius complies the schema and registers the custom resource type Plaid in UCP

    1. Deb lists the resources types he has in his environment

        ```bash
        rad resource-provider list -e myenv
        ```
        Radius lists the custom resource types that Deb has registered in his environment

2. Amy is a system integrator who is familiar with OpenAPI schema format and wants to author a custom resource PostgreSQL in Radius

    1. Amy uses the radCLI to scaffold a yaml template definition

        ```bash

>         rad resource-provider init plaid --template-kind yaml

Reviewer 5: should this be:
 rad resource-provider init Applications --template-kind yaml
 since Applications is the rp used below?

Author: yes

>         rad resource-provider init plaid --template-kind yaml

Author: ```suggestion
        rad resource-provider init Applications --template-kind yaml
```

        ```
        A sample yaml file is created in the application folder.

    1. Amy authors the schema for plaid in typespec format
        
        ```
        namespace: 'Applications.Datastores'
        apiVersion: 'v1'
        scope: '/planes/radius/Applications.Datastores'
        resourceTypes:
            'Applications.Datastores/postgreSQL':
                name: 'Name of the resource'
                description: 'PostgreSQL' 
                properties: 
                required: ['host', 'env', 'secrets']
                host:
                    type: 'string'
                    description: 'Hostname'
                env:
                    type: 'string'
                    description: 'Radius environment'
                app:
                    type: 'string'
                    description: 'Radius application'
                secrets:
                    connectionstring:
                    type: 'string'
                    description: 'Connection string for PostgreSQL'
        ```

>     1. Amy adds a version for the schema

Reviewer 6: Does the yaml ultimately end up being compiled into the same format as the output as the TSP compiler? Or are we going to support two formats?

Author: I think we shd merge to the open API format typespec compiles to. Ultimately we want to be users of our own solution and refactor the Radius built in types to use UDT

        scope: '/planes/radius/Applications.Datastores'
        resourceTypes:
            'Applications.Datastores/postgreSQL':
                name: 'Name of the resource'
                description: 'PostgreSQL' 
                properties: 
                required: ['host', 'env', 'secrets']
                host:
                    type: 'string'
                    description: 'Hostname'
                env:
                    type: 'string'
                    description: 'Radius environment'
                app:
                    type: 'string'
                    description: 'Radius application'
                secrets:
                    connectionstring:
                    type: 'string'
                    description: 'Connection string for PostgreSQL'
        ```
   1. Amy registers the postgreSQL schema to an existing provider in Radius
    
        ```bash
        rad resource-type add -rp Applications --template-kind yaml -e myenv -v v1
        ```
        
        Radius complies the schema and registers the custom resource type Plaid in UCP
        
        Question : Will we continue to accept new types like postgreSQL to `Applications` namespace

>    

Reviewer: BTW I feel like this is a great way place to talk about community supported assets like a "contrib" repository of resource types. Now that we've explained the work to define one we can talk about the goal of reuse with more clarity.

#### Iterating and testing the application platform

 As a platform engineer, I need a productive inner loop for development and testing the application platform. Simple and easy way to iterate on changes, test and deploy them

**Solution in Radius**

1. Bicep tooling support to test and deploy the custom resource types - rad CLI support to generate and publish the Bicep types for the custom resources to the registry.
1. Documentation to guide the users on how to test and deploy the custom resource types in Radius

*User groups and their needs*

There are two types of users that we need to cater to:
1. Users who have registered schema using typespec schema format
1. Users who have registered schema using yaml schema format

*User experience*

1. Deb compiles the typespec schema 
    
        ```bash
        tsp compile
        ```
        The typespec schema is compiled and validated for any errors and warnings

1. Deb publishes the bicep types to a registry
    
        ```bash

>         rad publish-bicep-types --file contoso.tsp --registry myregistry

Reviewer 3: Should we be keeping track of versioning here as well? the index.json file would have an extension version when it gets generated so that should match with the API version

        name: plaidName
        description: plaidDescription
        queue: plaidQueue
    }
    ```
1. Deb uses the rad CLI to deploy the application with the custom resource type Plaid

    ```bash
    rad deploy app.bicep 
    ```
    The application is deployed with the custom resource type Plaid in Radius. Deb can iterate on the changes, compile and test the application with the custom resource type.

#### Ship the custom platform

As a platform engineer, I want to register the API's to the platform, provide documentation, onboarding and support. Intuitive, seamless, collaborative

**Solution in Radius**

1. End-end-documentation for the custom resource types - Radius provides documentation to the users on how to author, test and deploy the custom resource types in Radius.
    1. Typespec samples and documentation - Provide samples and documentation for the users to author the schema in typespec format
    1. Documentation for generating Radius-Bicep types - Reuse existing pipelines and GH workflows used in Radius to generate Radius types in Bicep as an example and provide documentation to the users on how to build and publish the custom resource types in Radius.

> Question - Documentation seems like a good starting point for users to get started on using custom resource types? What are other ways, should we provide sdk's to author the schema and generate the types?

Reviewer: I think documentation and samples is the default starting answer and required for anything we build. It also feels like we're already signing up to build some pretty rich tooling (shown in the design here). Let's discuss if you think we should do more.

Author: template repositoryry, scaffolding template, tooling

#### Maintain the custom platform

As a platform engineer, I want to version and maintain the API's. Ease of support, sense of ownership (decision making)

**Solution in Radius**

* Versioning of the API's

> * Versioning of the bicep types

Reviewer 3: Should the versioning of the API correspond with the version of the bicep types used in the `bicepconfig.json`?

- Developers : Developers are responsible for building the cloud native applications. They are responsible for writing the code, designing and maintaining the applications.

- Open-source contributors : Cloud native open-source contributors can be any of the above personas who are interested in contributing to the cloud native projects. 

### Positive user outcomes

- Author and deploy : I can bring my own applications with custom resources/services and integrate with Radius seamlessly. I don’t have to write Go code to integrate with Radius components and can use simple resource specification to generate the resource definition and integrate with Radius 

- Recipes for user defined types: I can create Recipes for the custom resource types and deploy via Radius 

- OSS contributions: I can create custom resource types and Recipes and open-source it for the community to discover and use.

## Key scenarios

### Scenario 1: Deb integrates the Budgets app with Radius

> Deb, a platform engineer at a Financial Services company who is responsible for building the custom internal developer platform. He has started the journey to enable radification one of their Budgets app. The Budgets app relies on an internal messaging service called Plaid. Deb wants to use Plaid as a resource type in Radius to deploy the application seamlessly. Deb needs a to author, register and deploy Plaid as a user defined type in Radius without having to write Go code to integrate with Radius components. This will enable his team, IT operators and application developers to leverage all the features of Radius, such as Recipes, Connections, and the rad CLI, with ease and flexibility.

Author: ```suggestion
Deb, a platform engineer at a Financial Services company who is responsible for building the custom internal developer platform. He has started the journey to enable radification one of their Budgets app. The Budgets app relies on an internal messaging service called Plaid. Deb wants to use Plaid as a resource type in Radius to deploy the application seamlessly and leverage all the features of Radius, such as Recipes, Connections, and the rad CLI, with ease and flexibility.
```

## Key assumptions to test and questions to answer
<!-- If you are making assumptions that, if incorrect, would cause us to significantly alter our approach to this scenario, make them explicit here.  Also call out how / when you plan to validate key assumptions. -->
<!-- What big questions must we answer in order to clarify our plan for this scenario.  When and how do you plan to answer those questions (prototype feature x, user research, competitive research, etc) -->

- Assumption: Users will find value in the extensibility feature and actively contribute to the community.
- Assumption: Users will use extensibility in lieu of existing extenders because they want to take advantage of things like strong typing and validation in Radius
- Assumption: Service providers of proprietary services will find value in creating and publishing Recipes for their services hoping to increase the adoption of their services.

## Current state 
<!-- If we already have some ongoing investment in this area, summarize the current state and point to any relevant documents. -->

### Current state in Radius

> [Radius Extenders](https://docs.radapp.io/guides/author-apps/custom/overview/#extenders) enables users to author custom resource types but they are weakly typed. They allow you to pass in any property or secret and for cases where the developer/operator need to extend and reference a simple resource type definition. Extenders are for untyped/ weakly typed resources and do not provide a lot of customization for users wanting to have guardrails in their platforms with strongly typed resource definitions following a strict set of rules and restrictions. 

Author: ```suggestion
[Radius Extenders](https://docs.radapp.io/guides/author-apps/custom/overview/#extenders) enables users to author custom resource types. They allow you to pass in any property or secret and for cases where the developer/operator need to extend and reference a simple resource type definition. Extenders are for untyped/ weakly typed resources and do not provide a lot of customization for users wanting to have guardrails in their platforms with strongly typed resource definitions following a strict set of rules and restrictions. 
```
